### PR TITLE
[Fix][Txn] Fix transaction redeliver message if ack failed

### PR DESCRIPTION
**Pull request by liangyepianzhou**
### Motivation & Modification
If transaction is not in the state of open, we should not allow it to ack message with in transaction. And the message should be redeliver again.
### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (yes / no)
  - The schema: (yes / no / don't know)
  - The default values of configurations: (yes / no)
  - The wire protocol: (yes / no)
  - The rest endpoints: (yes / no)
  - The admin cli options: (yes / no)
  - Anything that affects deployment: (yes / no / don't know)

### Documentation

Check the box below or label this PR directly.

Need to update docs? 

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)
  
- [x] `no-need-doc` 
(Please explain why)
  
- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-added`
(Docs have been already added)

### CODE REVIEW

**Review by codelipenghui — APPROVED** [review]

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/TransactionEndToEndTest.java [hunk 1]
```diff
@@ -1052,13 +1052,16 @@ public void testTxnTimeOutInClient() throws Exception{
             Assert.assertTrue(e.getCause().getCause() instanceof TransactionCoordinatorClientException
                     .InvalidTxnStatusException);
         }
+        Message<String> message = null;
         try {
-            Message<String> message = consumer.receive();
+            message = consumer.receive();
```
Comment (mattisonchao) [inline pulsar-broker/src/test/java/org/apache/pulsar/client/impl/TransactionEndToEndTest.java:1057]: It feels like adding a timeout to the method avoids getting blocks all the time. But I'm not sure if we have another mechanism to guarantee that.

**Review by mattisonchao — APPROVED** [review]
LGTM, Left a small comment.

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/TransactionEndToEndTest.java (hunk 1 above)
Comment (liangyepianzhou) [reply pulsar-broker/src/test/java/org/apache/pulsar/client/impl/TransactionEndToEndTest.java:1057]: Ci test will run for up to 30s, and an error will be reported if it times out. Of course, we cannot rely on this.

**Review by gaoran10 — APPROVED** [review]
LGTM

**Comment by Technoboy-** [comment]
The test only tests for 1 consumer, and only tests redelivery. 
What's the behavior when there are 2 or more consumers?

**Comment by liangyepianzhou** [comment]
> The test only tests for 1 consumer, and only tests redelivery. What's the behavior when there are 2 or more consumers?

What we need to test is that it does call the `redeliverUnacknowledgedMessages` method. As for whether this method fully meets expectations, it should be tested by the PR that writes this method. For details, you can see the `ConsumerRedeliveryTest`, `DispatcherBlockConsumerTest` and `SimpleProducerConsumerTest

**Review by HQebupt — APPROVED** [review]

**Comment by codelipenghui** [comment]
https://github.com/apache/pulsar/pull/15593 revert the PR, so don't need to cherry-pick to release branches

**Pull request merged**